### ISSUE TITLE
Incorrect Program Version (But with a twist)

### ISSUE THREAD

**Issue by ThioJoe** [microsoft/winget-cli#3506]
### Brief description of your issue

**TL;DR:** WinGet thinks the installed version of [Sonarr](https://github.com/microsoft/winget-pkgs/tree/master/manifests/t/TeamSonarr/Sonarr) is 3.0.9.1549, not the actually-installed latest 3.0.10.1567, _despite the former version never being installed on the system, nor it being the version appearing in Add/Remove Programs._

-------------------

**Full Long-winded explanation:**

I know the usual causes for incorrect package versions, and that WinGet theoretically gets the version number from the same place the Add/Remove Programs list does. And that it also gets it from the 'DisplayVersion' registry key for the installation.

But in this situation WinGet is displaying the installed version of [Sonarr](https://github.com/microsoft/winget-pkgs/tree/master/manifests/t/TeamSonarr/Sonarr) as being version 3.0.9.1549, when it is actually 3.0.10.1567. It does correctly list the latest available as 3.0.10.1567.

A few things are confusing me enough to submit this as a bug report though:

- I've never had version 3.0.9.1549 installed on this computer. It's literally a clean install of Windows from 2 weeks ago and I installed 3.0.10.1567 from the get-go.
- I did not even install Sonarr using WinGet, but their rather their installer exe
- The 'DisplayVersion' in the registry actually lists "3.0" as the version, which is also what Add/Remove Programs shows. So I'm not actually sure where it is even getting the 3.0.9.1549 value from.
- I opened up the index.db database and it indeed lists version 3.0.9.1549 corresponding to the installed version of Sonarr. So I assume that is where it's pulling the version from, I just don't know how it got that number in the first place.

----------------------

**Main questions:**
- What other places besides the registry does WinGet look for a version number?

-----------------

### Steps to reproduce

Run `winget list` or `winget upgrade`

### Expected behavior

It would show the correct 3.0.10.1567 installed version when doing `winget list` or `winget upgrade`, and would not show an update available.

### Actual behavior

It shows 
```
Name                                                     Id                           Version       Available     Source
------------------------------------------------------------------------------------------------------------------------
Sonarr version 3.0                                       TeamSonarr.Sonarr            3.0.9.1549    3.0.10.1567   winget
```

### Environment

```shell
Windows Package Manager (Preview) v1.6.1573-preview
Copyright (c) Microsoft Corporation. All rights reserved.

Windows: Windows.Desktop v10.0.22621.2070
System Architecture: X64
Package: Microsoft.DesktopAppInstaller v1.21.1573.0

Winget Directories
-------------------------------------------------------------------------------------------------------------------------------
Logs                               %LOCALAPPDATA%\Packages\Microsoft.DesktopAppInstaller_8wekyb3d8bbwe\LocalState\DiagOutputDir
User Settings                      %LOCALAPPDATA%\Packages\Microsoft.DesktopAppInstaller_8wekyb3d8bbwe\LocalState\settings.json
Portable Links Directory (User)    %LOCALAPPDATA%\Microsoft\WinGet\Links
Portable Links Directory (Machine) C:\Program Files\WinGet\Links
Portable Package Root (User)       %LOCALAPPDATA%\Microsoft\WinGet\Packages
Portable Package Root              C:\Program Files\WinGet\Packages
Portable Package Root (x86)        C:\Program Files (x86)\WinGet\Packages

Links
---------------------------------------------------------------------------
Privacy Statement   https://aka.ms/winget-privacy
License Agreement   https://aka.ms/winget-license
Third Party Notices https://aka.ms/winget-3rdPartyNotice
Homepage            https://aka.ms/winget
Windows Store Terms https://www.microsoft.com/en-us/storedocs/terms-of-sale

Admin Setting                             State
--------------------------------------------------
LocalManifestFiles                        Disabled
BypassCertificatePinningForMicrosoftStore Disabled
InstallerHashOverride                     Disabled
LocalArchiveMalwareScanOverride           Disabled
```

#### Log file for 'winget list sonarr'
```
2023-08-06 19:05:20.477 [CORE] WinGet, version [1.6.1573-preview], activity [{06EB9334-1B47-4D6C-A757-C156B967685A}]
2023-08-06 19:05:20.477 [CORE] OS: Windows.Desktop v10.0.22621.2070
2023-08-06 19:05:20.477 [CORE] Command line Args: winget  list sonarr
2023-08-06 19:05:20.477 [CORE] Package: Microsoft.DesktopAppInstaller v1.21.1573.0
2023-08-06 19:05:20.477 [CORE] IsCOMCall:0; Caller: winget-cli
2023-08-06 19:05:20.479 [CLI ] WinGet invoked with arguments: 'list' 'sonarr'
2023-08-06 19:05:20.479 [CLI ] Found subcommand: list
2023-08-06 19:05:20.479 [CLI ] Leaf command to execute: root:list
2023-08-06 19:05:20.480 [CLI ] Executing command: list
2023-08-06 19:05:20.487 [REPO] Default source requested, multiple sources available, adding all to source references.
2023-08-06 19:05:20.487 [REPO] Adding to source references msstore
2023-08-06 19:05:20.487 [REPO] Adding to source references winget
2023-08-06 19:05:20.488 [REPO] Multiple sources available, creating aggregated source.
2023-08-06 19:05:20.488 [REPO] Adding to aggregated source: msstore
2023-08-06 19:05:20.488 [REPO] Sending http GET request to: https://storeedgefd.dsx.mp.microsoft.com/v9.0/information
2023-08-06 19:05:20.567 [REPO] Response status: 200
2023-08-06 19:05:20.567 [REPO] Sending http GET request to: https://storeedgefd.dsx.mp.microsoft.com/v9.0/information
2023-08-06 19:05:20.584 [REPO] Response status: 200
2023-08-06 19:05:20.585 [REPO] Adding to aggregated source: winget
2023-08-06 19:05:20.589 [CORE] Examining extension: PFN = Microsoft.Winget.Source_8wekyb3d8bbwe, ID = IndexDB
2023-08-06 19:05:20.589 [CORE] Found matching extension.
2023-08-06 19:05:20.601 [REPO] Opening SQLite Index for ImmutableRead at 'C:\Program Files\WindowsApps\Microsoft.Winget.Source_2023.806.2141.619_neutral__8wekyb3d8bbwe\Public\index.db'
2023-08-06 19:05:20.601 [SQL ] Opening SQLite connection #1: 'C:\Program Files\WindowsApps\Microsoft.Winget.Source_2023.806.2141.619_neutral__8wekyb3d8bbwe\Public\index.db' [1, 40]
2023-08-06 19:05:20.601 [REPO] Opened SQLite Index with version [1.6], last write [2023-08-06 13:40:32.000]
2023-08-06 19:05:20.865 [REPO] Creating PredefinedInstalledSource with filter [None]
2023-08-06 19:05:20.865 [REPO] Creating new SQLite Index with version [Latest] at ':memory:'
2023-08-06 19:05:20.865 [SQL ] Opening SQLite connection #2: ':memory:' [6, 0]
2023-08-06 19:05:20.884 [REPO] Reading MSI UpgradeCodes
2023-08-06 19:05:20.885 [REPO] Examining ARP entries for Machine | X64
2023-08-06 19:05:20.899 [REPO] Examining ARP entries for Machine | X86
2023-08-06 19:05:20.916 [REPO] Reading MSI UpgradeCodes
2023-08-06 19:05:20.917 [REPO] Examining ARP entries for User | X64
2023-08-06 19:05:21.229 [REPO] Finding available package from installed package using system reference search: Query:[none] Include:ProductCode='{56c1065d-3523-4025-b76d-6f73f67f7f71}_is1'[Exact] Include:NormalizedNameAndPublisher='sonarr'+'teamsonarr'[Exact]
2023-08-06 19:05:21.232 [REPO] Opening SQLite Index for ReadWrite at 'C:\Users\Joe\AppData\Local\Packages\Microsoft.DesktopAppInstaller_8wekyb3d8bbwe\LocalState\StoreEdgeFD\installed.db'
2023-08-06 19:05:21.232 [SQL ] Opening SQLite connection #3: 'C:\Users\Joe\AppData\Local\Packages\Microsoft.DesktopAppInstaller_8wekyb3d8bbwe\LocalState\StoreEdgeFD\installed.db' [2, 0]
2023-08-06 19:05:21.232 [REPO] Opened SQLite Index with version [1.3], last write [2023-08-01 11:21:33.000]
2023-08-06 19:05:21.252 [REPO] Opening SQLite Index for ReadWrite at 'C:\Users\Joe\AppData\Local\Packages\Microsoft.DesktopAppInstaller_8wekyb3d8bbwe\LocalState\Microsoft.Winget.Source_8wekyb3d8bbwe\installed.db'
2023-08-06 19:05:21.252 [SQL ] Opening SQLite connection #4: 'C:\Users\Joe\AppData\Local\Packages\Microsoft.DesktopAppInstaller_8wekyb3d8bbwe\LocalState\Microsoft.Winget.Source_8wekyb3d8bbwe\installed.db' [2, 0]
2023-08-06 19:05:21.253 [REPO] Opened SQLite Index with version [1.6], last write [2023-08-06 16:27:38.000]
2023-08-06 19:05:21.274 [REPO] Sending http POST request to: https://storeedgefd.dsx.mp.microsoft.com/v9.0/manifestSearch
2023-08-06 19:05:21.942 [REPO] Response status: 200
2023-08-06 19:05:22.031 [REPO] Finding installed package from available package using system reference search: Query:[none] Include:NormalizedNameAndPublisher='radarr'+'teamradarr'[Exact]
2023-08-06 19:05:22.035 [REPO] Finding installed package from available package using system reference search: Query:[none] Include:NormalizedNameAndPublisher='lidarr'+'teamlidarr'[Exact]
2023-08-06 19:05:22.036 [REPO] Finding installed package from available package using system reference search: Query:[none] Include:NormalizedNameAndPublisher='bazarr'+'morpheus'[Exact]
2023-08-06 19:05:22.263 [REPO] Finding installed package from available package using system reference search: Query:[none] Include:NormalizedNameAndPublisher='jackett'+'jackett'[Exact]
2023-08-06 19:05:22.266 [REPO] Opening pinning index
2023-08-06 19:05:22.266 [REPO] Opening existing pinning index
2023-08-06 19:05:22.266 [REPO] Opening SQLite Index for ReadWrite at 'C:\Users\Joe\AppData\Local\Packages\Microsoft.DesktopAppInstaller_8wekyb3d8bbwe\LocalState\pinning.db'
2023-08-06 19:05:22.266 [SQL ] Opening SQLite connection #5: 'C:\Users\Joe\AppData\Local\Packages\Microsoft.DesktopAppInstaller_8wekyb3d8bbwe\LocalState\pinning.db' [2, 0]
2023-08-06 19:05:22.266 [REPO] Opened Pinning Index with version [1.0], last write [2023-07-28 09:08:14.000]
2023-08-06 19:05:22.274 [CLI ] Leaf command succeeded: root:list
```

#### Screenshots:

<img width="215" alt="image" src="https://github.com/microsoft/winget-cli/assets/12518330/fdc30bd3-5710-49b4-8349-07e8b666e850">
<img width="707" alt="image" src="https://github.com/microsoft/winget-cli/assets/12518330/2c45e59c-74df-430e-8827-50f1c47a4394">
<img width="502" alt="image" src="https://github.com/microsoft/winget-cli/assets/12518330/cb810f72-bff1-4d90-a559-a563bbd5ecf5">
<img width="485" alt="image" src="https://github.com/microsoft/winget-cli/assets/12518330/5aaff464-45d6-425e-9fcf-57af8c0d0412">

**Comment by Trenly** [comment]
> I believe you nailed it @mdanish-kh
> 
> @Trenly would your suggestion to improve marketing versions address this scenario?
> 
> * [Separate "PackageVersion" from "Marketing Version" #3061](https://github.com/microsoft/winget-cli/issues/3061)

I don't think it would because the publisher is writing bad data to the ARP entry, so there would still be no way to disambiguate between versions, even if the package has a semantically correct version.  The marketing versions would address if the ARP version was non-semantic, had incorrect sorting, etc. but it can't correct for when the publisher writes the same version into ARP for every version of the app

**Comment by mdanish-kh** [comment]
This should be resolved now.
<img src=https://github.com/microsoft/winget-cli/assets/88161975/c0d1b8d3-a467-4230-b33a-1003590c85f4 height=120>

**Comment by denelon** [comment]
@ThioJoe thanks for the detailed write up. This looks like one of those cases where the publisher needs to make some changes to their package/installer so the metadata can be aligned with manifests to provide the best experience.

**Comment by denelon** [comment]
@Trenly has designs on an improved flow for versions. I was keeping this work item open as a reminder to myself for a better way to handle them. I'm going to go ahead and close this issue now.
